### PR TITLE
bpo-1635741: Port mmap module to multiphase initialization

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue1635741

### CODE REVIEW

**Review by shihai1991 — APPROVED** [review]
LGTM ;)

File: Modules/mmapmodule.c [hunk 1]
```diff
@@ -1659,6 +1643,23 @@ PyInit_mmap(void)
     setint(dict, "MADV_PROTECT", MADV_PROTECT);
 #endif
 #endif // HAVE_MADVISE
```
Comment (vstinner) [inline Modules/mmapmodule.c:1651]: Please add something like `if (PyErr_Occurred()) { return -1; }`. setint() silently ignore errors, but can "leak" an exception.

setint() doesn't check for PyDict_SetItemString() failure, and it doesn't report errors :-(

File: Modules/mmapmodule.c [hunk 2]
```diff
@@ -1521,33 +1521,17 @@ setint(PyObject *d, const char *name, long value)
...
-        return NULL;
-    dict = PyModule_GetDict(module);
-    if (!dict)
-        return NULL;
+    PyObject *dict = PyModule_GetDict(module);
+    if (!dict) {
+        return -1;
+    }
     PyDict_SetItemString(dict, "error", PyExc_OSError);
     PyDict_SetItemString(dict, "mmap", (PyObject*) &mmap_object_type);
```
Comment (vstinner) [inline Modules/mmapmodule.c]: Maybe PyModule_AddType() could be used here.

**Review by vstinner — APPROVED** [review]
LGTM. mmap code would deserve better error handling, but that can be addressed in a separated PR.

Apart static types, I don't anything which would be shared by multiple module instances/interpreters.

File: Modules/mmapmodule.c [hunk 3]
```diff
@@ -1509,157 +1509,161 @@ new_mmap_object(PyTypeObject *type, PyObject *args, PyObject *kwdict)
...
-    NULL,
-    NULL,
-    NULL
-};
-
-PyMODINIT_FUNC
-PyInit_mmap(void)
-{
-    PyObject *dict, *module;
+    if (PyModule_AddObject(module, "error", PyExc_OSError) < 0) {
```
Comment (vstinner) [inline Modules/mmapmodule.c:1520]: You need Py_INCREF before, and Py_DECREF on error.

File: Modules/mmapmodule.c [hunk 4]
```diff
@@ -1509,157 +1509,161 @@ new_mmap_object(PyTypeObject *type, PyObject *args, PyObject *kwdict)
...
+        return -1;
+    }
+    if (PyModule_AddType(module, &mmap_object_type) < 0) {
+        return -1;
+    }
 
-    if (PyType_Ready(&mmap_object_type) < 0)
-        return NULL;
+#define ADD_INT_MACRO(module, macro)                                  \
```
Comment (vstinner) [inline Modules/mmapmodule.c]: I suggest to rename the second parameter to "constant" rather than "macro".

**Review by vstinner — APPROVED** [review]
LGTM apart the PyExc_OSError bug.

**Comment by corona10** [comment]
@vstinner @shihai1991 Thanks for the reviews!

**Pull request merged**